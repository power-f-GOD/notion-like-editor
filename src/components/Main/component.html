<main class="flex flex-col container mx-auto max-w-3xl p-4 lg:px-0">
  <header
    class="flex gap-1 p-1 px-1.5 items-center justify-between border border-solid border-black/10 shadow-sm rounded-md text-sm">
    <section class="flex items-center gap-2.5">
      <span class="inline-block text-sm py-1 px-2 text-green-900 bg-green-100 rounded-md font-semibold">P</span>
      <span class="opacity-20">|</span>
      <span class="inline-flex items-center gap-1 opacity-50">%clock% Omin</span>
      <span class="opacity-20">|</span>
      <span class="w-5 h-5 rounded-full bg-orange-500">
        <!-- <img src="" alt=""> -->
      </span>
      <span class="opacity-20">|</span>
      <span class="inline-flex items-center gap-1 opacity-50">%arrow-down-left% 0</span>
    </section>

    <section class="flex gap-1.5 items-center">
      <span class="inline-flex opacity-50">%check-circle%</span>
      <span class="inline-flex text-green-700">%cloud%</span>
      <button class="px-0.5 py-1" tabindex="-1">%dots-three-vertical%</button>
    </section>
  </header>

  <section>
    <h1 class="text-[2.5rem] pb-2 mt-7 mb-5 border-b border-0 border-solid border-black/10">Front-end
      developer test
      project
    </h1>

    <p class="opacity-80">
      Your goal is to make a page that looks exactly like this one, and has the ability to create H1 text simply by
      typing / then 1, then typing text, and hitting Enter
    </p>
  </section>

  <section class="relative mt-10" data-textareas-container></section>
</main>