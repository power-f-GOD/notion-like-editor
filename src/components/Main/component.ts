import { SVGIconName } from 'src/types';
import { preprocess } from 'src/utils';
import { SVGIcon } from '../SVGIcon';
import mainHtml from './component.html?raw';
import textareaHtml from './textarea.html?raw';

export const Main = () => {
  setTimeout(() => {
    console.log('Main component mounted!');
  });

  return preprocess<SVGIconName>(mainHtml, {
    clock: SVGIcon('clock', 'w-4 h-4 min-w-[1rem]'),
    'arrow-down-left': SVGIcon('arrow-down-left'),
    'check-circle': SVGIcon('check-circle'),
    cloud: SVGIcon('cloud'),
    'dots-three-vertical': SVGIcon('dots-three-vertical')
  });
};

export const getTextareaTemplate = (props: {
  id: string;
  tag: string;
  placeholder: string;
}) => {
  return preprocess(textareaHtml, props);
};
