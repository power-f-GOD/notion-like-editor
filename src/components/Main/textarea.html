<%tag% class="relative left-0 top-0 z-10 mb-2 ring-none outline-none whitespace-prewrap" contenteditable="true"
  id="%id%" data-textarea placeholder="%placeholder%"></%tag%>