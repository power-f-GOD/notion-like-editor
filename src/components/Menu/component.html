<div class="menu-overlay show-menu fixed top-0 lef-0 z-[1000] overflow-hidden pointer-events-none">
  <div class="pointer-events-auto relative">
    <div class="fixed top-0 left-0" data-caret-observer>
      <span class="relative inline-block w-0 h-6"></span>
      <div class="w-80 h-96 absolute left-0 top-full bg-white z-100 max-w-[calc(100vw-2rem)] max-h-[40vh]" role="menu">
        <section
          class="w-full h-full relative top-0 py-3 border border-solid border-black/10 shadow-2xl rounded-lg bg-white overflow-auto">
          <h2 class="text-base px-3 font-semibold">Add blocks</h2>
          <p class="px-3 opacity-40 text-sm mb-3">Keep typing to filter, or escape to exit...</p>
          <p class="px-3 opacity-60 text-sm">
            Filtering keyword
            <span class='bg-blue-900 text-white rounded-sm px-1 py-0.5' data-feedback>
              /
            </span>
          </p>
          <ul class="list-none p-0 my-3"></ul>
        </section>
      </div>
    </div>
  </div>
</div>