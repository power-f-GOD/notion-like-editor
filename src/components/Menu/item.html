<li>
  <button class="flex text-sm px-3 py-1 items-center justify-start gap-2 rounded-none w-full" data-tag="%tag%"
    data-placeholder="%placeholder%">
    <span class="py-0.5 inline-flex">%text-t%</span>
    <span class="flex flex-col items-start">
      <strong class="leading-5">%title%</strong>
      <small class="opacity-50 leading-4">%description%</small>
    </span>
  </button>
</li>