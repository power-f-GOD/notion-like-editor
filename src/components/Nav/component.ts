import { Component, SVGIconName } from 'src/types';
import { preprocess } from 'src/utils';
import { SVGIcon } from '../SVGIcon';
import html from './component.html?raw';

export class Nav<
  Variables extends SVGIconName = SVGIconName
> extends Component<Variables> {
  init(): void {
    console.log('Nav component mounted!');
  }
}

export const nav = new Nav(
  {
    html,
    variables: {
      'caret-double-right': SVGIcon(
        'caret-double-right',
        'w-4 h-4 min-w-[1rem]'
      ),
      'book-open': SVGIcon('book-open'),
      'caret-down': SVGIcon('caret-down'),
      'lock-key-open': SVGIcon('lock-key-open')
    }
  },
  preprocess
);
