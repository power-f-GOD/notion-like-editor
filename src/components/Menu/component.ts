import { preprocess } from 'src/utils';
import { SVGIcon } from '../SVGIcon';
import { Main, main } from '../Main/component';
import { Component } from 'src/types';
import html from './component.html?raw';
import itemHtml from './item.html?raw';

export class Menu extends Component {
  private caretObserver!: HTMLElement;
  private menu!: HTMLElement;
  private list!: HTMLUListElement;
  private feedback!: HTMLElement;
  private caretRect: Partial<DOMRect> | null = null;

  private blocks: Array<{
    title: string;
    tag: 'p' | 'h1' | 'h2' | 'h3';
    description: string;
    placeholder?: string;
  }> = [
    {
      title: 'Text',
      tag: 'p',
      description: 'Plain text or paragraph.',
      placeholder:
        "Type '/' for block, ('@' to link docs or people doesn't work for nowðŸ˜Œ)"
    },
    { title: 'Heading 1', tag: 'h1', description: 'Big section heading.' },
    { title: 'Heading 2', tag: 'h2', description: 'Medium section heading.' },
    { title: 'Heading 3', tag: 'h3', description: 'Small section heading.' }
  ];

  constructor(
    protected template: (typeof Component)['prototype']['template'],
    protected preprocessor: typeof preprocess,
    private mainComponent: Main,
    private icon: typeof SVGIcon
  ) {
    super(template, preprocessor);
  }

  init() {
    console.log('Menu component mounted!');
    this.caretObserver = this.query('[data-caret-observer]')!;
    this.menu = this.caretObserver.querySelector(
      "[role='menu']"
    ) as HTMLElement;
    this.list = this.menu.querySelector('ul')!;
    this.feedback = this.menu.querySelector('[data-feedback]')!;
    this.appendTextarea();
    document.body.addEventListener('click', (e) => {
      if (!this.menu.contains(e.target as HTMLElement)) {
        this.showMenu(false);
      }
    });
  }

  private showMenu(show: boolean) {
    this.caretObserver.dataset.showmenu = String(show);
  }

  private moveTextareaCaretToPosition = (textarea: HTMLElement | Element) => {
    const range = document.createRange();
    const selection = window.getSelection();
    const prevOffset = +((textarea as HTMLElement).dataset.offset || -1);
    const hasOffset = prevOffset > -1;

    if (hasOffset) range.setStart(textarea.childNodes[0], prevOffset);
    range.collapse(hasOffset);
    selection!.removeAllRanges();
    selection!.addRange(range);
    if (!hasOffset) selection?.collapseToEnd();
    (textarea as HTMLElement).focus();
  };

  private switchContextTextarea = (e: KeyboardEvent, switchable?: boolean) => {
    const textarea = e.currentTarget as HTMLElement;
    const selection = window.getSelection();
    const range = selection && selection?.getRangeAt(0);
    const caretRect = range?.getBoundingClientRect();
    const textareaRect = textarea.getBoundingClientRect();
    const {
      previousElementSibling: previousTextarea,
      nextElementSibling: nextTextarea
    } = textarea;
    const textLength = textarea.textContent?.length || 0;

    // Save current caret offset/position, so to be remembered when re-focused
    textarea.dataset.offset = String(range?.startOffset || textLength - 1);

    if (switchable) {
      const isEmpty = !textLength;

      if (
        (e.key === 'ArrowUp' ||
          e.key === 'ArrowLeft' ||
          e.key === 'Backspace') &&
        previousTextarea
      ) {
        const isAtStart = range?.startOffset === 0;

        if (e.key === 'Backspace' && isEmpty && isAtStart) {
          e.preventDefault();
          textarea.parentNode?.removeChild(textarea);
        }

        if (
          e.key === 'ArrowUp' ||
          (e.key === 'ArrowLeft' && isAtStart) ||
          isEmpty
        ) {
          e.preventDefault();
          this.moveTextareaCaretToPosition(previousTextarea);
        }
      } else if (
        (e.key === 'ArrowDown' || e.key === 'ArrowRight') &&
        nextTextarea
      ) {
        if (range?.startOffset === textLength || e.key === 'ArrowDown') {
          e.preventDefault();
          this.moveTextareaCaretToPosition(nextTextarea);
        }
      } else if (e.key === 'Enter') {
        e.preventDefault();
        // We don't want duplicate textareas to be added when menu is displaying as the click listener on the menu adds that for us
        this.appendTextarea(
          textarea,
          this.list.querySelector('button')?.dataset as
            | { tag: string; placeholder: string }
            | undefined
        );
        this.showMenu(false);
      }
    }

    return { selection, range, caretRect, textareaRect, textarea };
  };

  private setMenuPosition = (e: KeyboardEvent) => {
    const { range, caretRect, textarea } = this.switchContextTextarea(e);
    const text = textarea.textContent!;
    const partialText = text
      .trim()
      .slice(
        Math.max((range?.startOffset || 0) - 8, 0),
        (range?.startOffset || 0) + 1
      );
    const canShowMenu = partialText.includes('/') && e.key !== 'Escape';

    // We don't want the menu moving as the user types, hence we check if it's currently displaying, i.e. bound would most likely be null at this point
    if (caretRect?.y && !this.caretRect) this.caretRect = caretRect;

    if (this.caretRect && canShowMenu) {
      const textFilter = partialText.replace(/.*\/\s*(\w*)/, '$1');
      const left =
        this.caretRect.left ||
        (textarea.offsetParent as HTMLElement).offsetLeft ||
        0;

      // Add menu list items
      this.list.innerHTML = '';
      this.feedback.textContent = textFilter || '/';
      this.blocks
        .filter(
          (block) =>
            partialText === '/' ||
            new RegExp(`${textFilter}`, 'i').test(block.title)
        )
        .forEach((block) => {
          this.list.insertAdjacentHTML(
            'beforeend',
            preprocess(itemHtml, {
              ...block,
              placeholder: block.placeholder || block.title,
              'text-t': this.icon('text-t', 'min-w-[1.75rem] opacity-40')
            })
          );
        });

      if (this.list.childElementCount === 0) {
        this.list.innerHTML = `<p class="p-3">No results.ðŸ¥²</p>`;
      }

      // Add a click listener to parent list element with which child button elements will, sort of, inherit from
      this.list.onclick = (e) => {
        const itemElement = e.target as HTMLElement;

        if (itemElement.tagName !== 'BUTTON') return;
        this.appendTextarea(
          textarea,
          itemElement.dataset as { tag: string; placeholder: string }
        );
        this.showMenu(false);
      };
      this.caretObserver.style.top = `${this.caretRect.top || 0}px`;
      this.caretObserver.style.left = `${left}px`;
      this.caretObserver.style.height = `${this.caretRect.height || 0}px`;

      // Move menu accordingly in x-axis if all part of it is not visible within the screen
      if (left + 336 > window.innerWidth) {
        this.menu.style.transformOrigin = `${window.innerWidth - left}px top`;
        this.menu.style.left = `${window.innerWidth - (left + 336)}px`;
      } else {
        this.menu.style.left = '0';
        this.menu.style.transformOrigin = 'left top';
      }

      // Move menu accordingly in y-axis if all part of it is not visible within the screen
      if (
        this.menu.offsetHeight >
        window.innerHeight - this.caretObserver.offsetTop - 32
      ) {
        this.menu.style.bottom = '100%';
        this.menu.style.top = 'unset';
        this.menu.style.transformOrigin =
          this.menu.style.transformOrigin.replace('top', 'bottom');
      } else {
        this.menu.style.bottom = 'unset';
        this.menu.style.top = '100%';
        this.menu.style.transformOrigin =
          this.menu.style.transformOrigin.replace('bottom', 'top');
      }
    }

    this.showMenu(canShowMenu);
    if (!canShowMenu) this.caretRect = null;
  };

  private appendTextarea = (
    node?: HTMLElement,
    props?: { tag: string; placeholder: string }
  ) => {
    const id = String(Math.random()).replace('0.', 'id');
    const template = this.mainComponent.getTextareaTemplate({
      id,
      tag: 'div',
      placeholder: this.blocks[0].placeholder!,
      ...props
    });
    const textareasContainer = this.query(
      '[data-textareas-container]'
    ) as HTMLElement;

    if (node) node.insertAdjacentHTML('afterend', template);
    else {
      textareasContainer.insertAdjacentHTML('beforeend', template);
    }

    const textarea = this.query(`#${id}`)!;

    textarea.addEventListener('keydown', (e) => {
      this.switchContextTextarea(e, true);
    });
    textarea.addEventListener('keyup', (e) => {
      if (e.key === 'Enter') {
        e.preventDefault();
      } else this.setMenuPosition(e as any);
    });
    textarea.addEventListener('blur', () => (this.caretRect = null));
    textarea.focus();
  };
}

export const menu = new Menu({ html }, preprocess, main, SVGIcon);
