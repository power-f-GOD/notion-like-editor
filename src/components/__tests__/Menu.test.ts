import '@testing-library/jest-dom';

import { menu } from '../Menu/component';
import html from '../Menu/component.html?raw';
import itemHtml from '../Menu/item.html?raw';
import { main } from '../Main/component';

describe('Menu', () => {
  it('renders the Menu component template unchanged', () => {
    expect(html).toMatchSnapshot();
  });

  it('renders the item template unchanged', () => {
    expect(itemHtml).toMatchSnapshot();
  });

  it('renders a textarea and the Menu component correctly', async () => {
    const body = document.body;

    body.insertAdjacentHTML('beforeend', main.render());
    expect(
      body.querySelector('[data-textareas-container]')
    ).toBeInTheDocument();
    body.insertAdjacentHTML('beforeend', menu.render());
    expect(body.innerHTML).toMatchSnapshot();
    await new Promise((resolve) => {
      setTimeout(resolve);
    });

    const textarea = body.querySelector(
      'div[contenteditable="true"]'
    ) as HTMLElement;

    expect(textarea).toBeInTheDocument();
    expect(textarea.innerHTML).toMatchSnapshot();
    expect(textarea.getAttribute('placeholder')).toBe(
      "Type '/' for block, ('@' to link docs or people doesn't work for nowðŸ˜Œ)"
    );
    expect(body.querySelector('div[role="menu"]')).toBeInTheDocument();
  });
});
