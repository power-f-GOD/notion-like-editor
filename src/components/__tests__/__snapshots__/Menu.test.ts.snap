// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Menu > renders a textarea and the Menu component correctly 1`] = `
"<main class=\\"flex flex-col container mx-auto max-w-3xl p-4 lg:px-0\\">
  <header class=\\"flex gap-1 p-1 px-1.5 items-center justify-between border border-solid border-black/10 shadow-sm rounded-md text-sm\\">
    <section class=\\"flex items-center gap-2.5\\">
      <span class=\\"inline-block text-sm py-1 px-2 text-green-900 bg-green-100 rounded-md font-semibold\\">P</span>
      <span class=\\"opacity-20\\">|</span>
      <span class=\\"inline-flex items-center gap-1 opacity-50\\"><svg class=\\"w-4 h-4 min-w-[1rem]\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 256 256\\"><rect width=\\"256\\" height=\\"256\\" fill=\\"none\\"></rect><circle cx=\\"128\\" cy=\\"128\\" r=\\"96\\" fill=\\"none\\" stroke=\\"currentColor\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"16\\"></circle><polyline points=\\"128 72 128 128 184 128\\" fill=\\"none\\" stroke=\\"currentColor\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"16\\"></polyline></svg> Omin</span>
      <span class=\\"opacity-20\\">|</span>
      <span class=\\"w-5 h-5 rounded-full bg-orange-500\\"></span>
      <span class=\\"opacity-20\\">|</span>
      <span class=\\"inline-flex items-center gap-1 opacity-50\\"><svg class=\\"w-5 h-5 min-w-[1.25rem]\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 256 256\\"><rect width=\\"256\\" height=\\"256\\" fill=\\"none\\"></rect><line x1=\\"192\\" y1=\\"64\\" x2=\\"64\\" y2=\\"192\\" fill=\\"none\\" stroke=\\"currentColor\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"16\\"></line><polyline points=\\"168 192 64 192 64 88\\" fill=\\"none\\" stroke=\\"currentColor\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"16\\"></polyline></svg> 0</span>
    </section>

    <section class=\\"flex gap-1.5 items-center\\">
      <span class=\\"inline-flex opacity-50\\"><svg class=\\"w-5 h-5 min-w-[1.25rem]\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 256 256\\"><rect width=\\"256\\" height=\\"256\\" fill=\\"none\\"></rect><polyline points=\\"88 136 112 160 168 104\\" fill=\\"none\\" stroke=\\"currentColor\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"16\\"></polyline><circle cx=\\"128\\" cy=\\"128\\" r=\\"96\\" fill=\\"none\\" stroke=\\"currentColor\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"16\\"></circle></svg></span>
      <span class=\\"inline-flex text-green-700\\"><svg class=\\"w-5 h-5 min-w-[1.25rem]\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 256 256\\"><rect width=\\"256\\" height=\\"256\\" fill=\\"none\\"></rect><path d=\\"M80,128a80,80,0,1,1,80,80H72A56,56,0,1,1,85.92,97.74\\" fill=\\"none\\" stroke=\\"currentColor\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"16\\"></path></svg><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 256 256\\"><rect width=\\"256\\" height=\\"256\\" fill=\\"none\\"></rect><path d=\\"M80,128a80,80,0,1,1,80,80H72A56,56,0,1,1,85.92,97.74\\" fill=\\"none\\" stroke=\\"currentColor\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"16\\"></path></svg></span>
      <button class=\\"px-0.5 py-1\\" tabindex=\\"-1\\"><svg class=\\"w-5 h-5 min-w-[1.25rem]\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 256 256\\"><rect width=\\"256\\" height=\\"256\\" fill=\\"none\\"></rect><circle cx=\\"128\\" cy=\\"128\\" r=\\"12\\"></circle><circle cx=\\"128\\" cy=\\"60\\" r=\\"12\\"></circle><circle cx=\\"128\\" cy=\\"196\\" r=\\"12\\"></circle></svg></button>
    </section>
  </header>

  <section>
    <h1 class=\\"text-[2.5rem] pb-2 mt-7 mb-5 border-b border-0 border-solid border-black/10\\">
      Front-end
      developer test
      project
    </h1>

    <p class=\\"opacity-80\\">
      Your goal is to make a page that looks exactly like this one, and has the ability to create H1 text simply by
      typing / then 1, then typing text, and hitting Enter
    </p>
  </section>

  <section class=\\"relative mt-10\\" data-textareas-container=\\"\\"></section>
</main><div class=\\"menu-overlay show-menu fixed top-0 lef-0 z-[1000] overflow-hidden pointer-events-none\\">
  <div class=\\"pointer-events-auto relative\\">
    <div class=\\"fixed top-0 left-0\\" data-caret-observer=\\"\\">
      <span class=\\"relative inline-block w-0 h-6\\"></span>
      <div class=\\"w-80 h-96 absolute left-0 top-full bg-white z-100 max-w-[calc(100vw-2rem)] max-h-[40vh]\\" role=\\"menu\\">
        <section class=\\"w-full h-full relative top-0 py-3 border border-solid border-black/10 shadow-2xl rounded-lg bg-white overflow-auto\\">
          <h2 class=\\"text-base px-3 font-semibold\\">Add blocks</h2>
          <p class=\\"px-3 opacity-40 text-sm mb-3\\">Keep typing to filter, or escape to exit...</p>
          <p class=\\"px-3 opacity-60 text-sm\\">
            Filtering keyword
            <span class=\\"bg-blue-900 text-white rounded-sm px-1 py-0.5\\" data-feedback=\\"\\">
              /
            </span>
          </p>
          <ul class=\\"list-none p-0 my-3\\"></ul>
        </section>
      </div>
    </div>
  </div>
</div>"
`;

exports[`Menu > renders a textarea and the Menu component correctly 2`] = `""`;

exports[`Menu > renders the Menu component template unchanged 1`] = `
"<div class=\\"menu-overlay show-menu fixed top-0 lef-0 z-[1000] overflow-hidden pointer-events-none\\">
  <div class=\\"pointer-events-auto relative\\">
    <div class=\\"fixed top-0 left-0\\" data-caret-observer>
      <span class=\\"relative inline-block w-0 h-6\\"></span>
      <div class=\\"w-80 h-96 absolute left-0 top-full bg-white z-100 max-w-[calc(100vw-2rem)] max-h-[40vh]\\" role=\\"menu\\">
        <section
          class=\\"w-full h-full relative top-0 py-3 border border-solid border-black/10 shadow-2xl rounded-lg bg-white overflow-auto\\">
          <h2 class=\\"text-base px-3 font-semibold\\">Add blocks</h2>
          <p class=\\"px-3 opacity-40 text-sm mb-3\\">Keep typing to filter, or escape to exit...</p>
          <p class=\\"px-3 opacity-60 text-sm\\">
            Filtering keyword
            <span class='bg-blue-900 text-white rounded-sm px-1 py-0.5' data-feedback>
              /
            </span>
          </p>
          <ul class=\\"list-none p-0 my-3\\"></ul>
        </section>
      </div>
    </div>
  </div>
</div>"
`;

exports[`Menu > renders the item template unchanged 1`] = `
"<li>
  <button class=\\"flex text-sm px-3 py-1 items-center justify-start gap-2 rounded-none w-full\\" data-tag=\\"%tag%\\"
    data-placeholder=\\"%placeholder%\\">
    <span class=\\"py-0.5 inline-flex\\">%text-t%</span>
    <span class=\\"flex flex-col items-start\\">
      <strong class=\\"leading-5\\">%title%</strong>
      <small class=\\"opacity-50 leading-4\\">%description%</small>
    </span>
  </button>
</li>"
`;
